### PR TITLE
8337066: Repeated call of StringBuffer.reverse with double byte string returns wrong result

### PR DESCRIPTION
**Pull request by veresov**
This is essentially a defensive forward port of a solution for an issue discovered in 11, where a use of pinned load (for which we disable almost all transforms in `LoadNode::Ideal()`) leads to a graph shape that is not expected by `insert_anti_dependences()`

The `insert_anti_dependences()` assumes that the memory input for a load is the root of the memory subgraph that it has to search for the possibly conflicting store. Usually this is true if we run all the memory optimizations but with pinned we don't.

Compare the good graph shape (with control dependency set to `UnknownControl`):
Good graph

With the graph produce with the nodes pinned:
Bad graph

With the "bad graph" loads and store don't share the same memory graph root, and therefore are not considered by `insert_anti_dependences()`. The solution, I think, could be to walk up the memory chain of the load, skipping MergeMems, in order to get to the real root and then run the precedence edge insertion algorithm from there.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8337066](https://bugs.openjdk.org/browse/JDK-8337066): Repeated call of StringBuffer.reverse with double byte string returns wrong result (**Bug** - P2)

### Reviewers
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**)
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**) Review applies to [e9295d93](https://git.openjdk.org/jdk/pull/21222/files/e9295d93c3efa4d989df3396cef7924f142fe9e2)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/21222/head:pull/21222` \
`$ git checkout pull/21222`

Update a local copy of the PR: \
`$ git checkout pull/21222` \
`$ git pull https://git.openjdk.org/jdk.git pull/21222/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 21222`

View PR using the GUI difftool: \
`$ git pr show -t 21222`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/21222.diff">https://git.openjdk.org/jdk/pull/21222.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/21222#issuecomment-2379622735)

### CODE REVIEW

File: src/hotspot/share/opto/gcm.cpp
Comment (vnkozlov) [inline src/hotspot/share/opto/gcm.cpp:753]: This will be executed for all loads and not only pinned. Is this okay?

Comment (veresov) [reply src/hotspot/share/opto/gcm.cpp:753]: Normally loads would not have a MergeMem as their memory input at this stage. `Ideal()` splits the memory, so that the memory input of a load is as precise as possible. So I think, that loop is benign for a regular load.

Comment (vnkozlov) [reply src/hotspot/share/opto/gcm.cpp:753]: Okay

**Review by vnkozlov — COMMENTED** [review]
Please, run additional testing I asked in JBS.

**Comment by veresov** [comment]
> Please, run additional testing I asked in JBS.

Good call. This showed a failure in one of the vectorization tests that is probably caused by this change. I need to take a further look.

**Comment by veresov** [comment]
> > Please, run additional testing I asked in JBS.
> 
> Good call. This showed a failure in one of the vectorization tests that is probably caused by this change. I need to take a further look.

I ran TestAlignVectorFuzzer a bunch of times during the weekend and this failure mode is preexisting. No caused by my change. So everything is clean.

**Review by vnkozlov — APPROVED** [review]
Good.

**Comment by vnkozlov** [comment]
You need second review for this

File: src/hotspot/share/opto/gcm.cpp
Comment (TobiHartmann) [reply src/hotspot/share/opto/gcm.cpp:753]: Would it still make sense to assert `load->control_dependency() == Pinned` here for now?

File: src/hotspot/share/opto/gcm.cpp
Comment (TobiHartmann) [inline src/hotspot/share/opto/gcm.cpp:750]:
```suggestion
  // We don't optimize the memory graph for pinned loads, so we may need to raise the
```

**Review by TobiHartmann — APPROVED** [review]
Looks good to me otherwise. You might want to run performance testing just to make sure.

**Comment by veresov** [comment]
> Looks good to me otherwise. You might want to run performance testing just to make sure.

While working on it I inserted a printf in it and the loop almost never happens since Loads are typically normalized. So, I don't think there is any impact on performance and I don't think checking for controlled dependency is necessary. It would also require us to carry this information to mach nodes...

**Comment by veresov** [comment]
Oh, I guess I need one of you guys to approve it again after I fixed the comment per Tobias' recommendation.

**Review by vnkozlov — APPROVED** [review]
Re-approved.

**Comment by veresov** [comment]
Thanks, Vladimir!

**Pull request closed without merging**
